contracts/core/MarginRequirements.sol

```diff
@@ -0,0 +1,185 @@
+
+    /**
+     * @notice sets the initial margin %
+     * @dev can only be called by owner
+     * @param _underlying underlying asset address
+     * @param _collateralAsset collateral asset address
+     * @param _isPut option type the vault is selling
+     * @param _account account address
+     * @param _initialMargin initial margin percentage (eg. 10% = 1000)
+     */
+    function setInitialMargin(
+        address _underlying,
+        address _collateralAsset,
+        bool _isPut,
+        address _account,
+        uint256 _initialMargin
+    ) external onlyOwner {
+        require(
+            _initialMargin > 0 && _initialMargin <= 100 * 10**2,
+            "MarginRequirements: initial margin cannot be 0 or higher than 100%"
+        );
+        require(_underlying != address(0), "MarginRequirements: invalid underlying");
+        require(_collateralAsset != address(0), "MarginRequirements: invalid collateral");
+        require(_account != address(0), "MarginRequirements: invalid account");
+
+        initialMargin[keccak256(abi.encode(_underlying, _collateralAsset, _isPut))][_account] = _initialMargin;
+    }
+
+    /**
+     * @notice sets the maintenance margin absolute amount
+     * @dev can only be called by keeper
+     * @param _vaultID id of the vault
+     * @param _maintenanceMargin maintenance margin absolute amount
+     */
+    function setMaintenanceMargin(uint256 _vaultID, uint256 _maintenanceMargin) external onlyKeeper {
+        require(_maintenanceMargin > 0, "MarginRequirements: initial margin cannot be 0");
+
+        maintenanceMargin[_vaultID] = _maintenanceMargin;
+    }
+
+    /************************************************
+     *  MARGIN OPERATIONS
+     ***********************************************/
+
+    /**
+     * @notice checks if there is enough collateral to mint the desired amount of otokens
+     * @param _account account address
+     * @param _notional order notional amount
+     * @param _underlying underlying asset address
+     * @param _isPut option type the vault is selling
+     * @param _collateralAsset collateral asset address
+     * @param _collateralAmount collateral amount
+     * @return boolean value stating whether there is enough collateral to mint
+     */
+    function checkMintCollateral(
+        address _account,
+        uint256 _notional,
+        address _underlying,
+        bool _isPut,
+        uint256 _collateralAmount,
+        address _collateralAsset
+    ) external view returns (bool) {
+        uint256 collateralDecimals = uint256(ERC20Interface(_collateralAsset).decimals());
+
+        uint256 initialMarginRequired = initialMargin[keccak256(abi.encode(_underlying, _collateralAsset, _isPut))][
+            _account
+        ];
+
+        return
+            _notional.mul(initialMarginRequired).mul(10**collateralDecimals).mul(1e8) <
+            _collateralAmount.mul(oracle.getPrice(_collateralAsset)).mul(100e2).mul(1e6);
+    }
+
```

Comment:
can you add in-line comments in checkMintCollateral, checkWithdrawCollateral that give high level overview of whats going on in the math

Reply:
https://github.com/ribbon-finance/GammaProtocol-OTC/pull/2/commits/2eac93d24b4e034056172eb761557d349317b04e